### PR TITLE
Add `RequiredDeep` type

### LINKED ISSUE #273: Proposal: `RequiredDeep`
similar to PartialDeep but the inverse, where all KV pairs are required.

### PR DESCRIPTION
**Pull request by ilchenkoArtem**
Fixes #273

### CODE REVIEW

**Comment by jmike** [comment]
note: the issue linked in the PR description points to the contribution guidelines instead of an actual issue. Is this by mistake?

**Comment by baoshan** [comment]
It’s [this issue](https://github.com/sindresorhus/type-fest/issues/273) I think.

**Comment by ilchenkoArtem** [comment]
> note: the issue linked in the PR description points to the contribution guidelines instead of an actual issue. Is this by mistake?

sorry(( wrong link

**Comment by ilchenkoArtem** [comment]
> It’s [this issue](https://github.com/sindresorhus/type-fest/issues/273) I think.

Yes you are right

File: source/required-deep.d.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	? E
+	: E extends Map<infer KeyType, infer ValueType>
+		? Map<RequiredDeep<KeyType>, RequiredDeep<ValueType>>
+		: E extends Set<infer ItemType>
+			? Set<RequiredDeep<ItemType>>
+			: E extends ReadonlyMap<infer KeyType, infer ValueType>
+				? ReadonlyMap<RequiredDeep<KeyType>, RequiredDeep<ValueType>>
+				: E extends ReadonlySet<infer ItemType>
+					? ReadonlySet<RequiredDeep<ItemType>>
+					: E extends (arg: any[]) => unknown
```
Comment (jarpoole) [inline source/required-deep.d.ts:50]: I think there is a small typo here, this type will only match  functions with a single array argument. From my (limited) testing it should actually be `: E extends (...arg: any[]) => unknown` to match all types of function args.

Comment (sindresorhus) [reply source/required-deep.d.ts:50]: @ilchenkoArtem Make sure you add a test for this too.

Comment (ilchenkoArtem) [reply source/required-deep.d.ts:50]: I wil fix this it this week

**Comment by sindresorhus** [comment]
Make sure you look at our existing "deep" types and that this new types handles all the cases handled by them.

For example, https://github.com/sindresorhus/type-fest/blob/2f418dbbb6182c53cdac26dc3421f7e8806c789b/source/readonly-deep.d.ts#L80

File: test-d/required-deep.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		readonlySet?: ReadonlySet<string | undefined>;
+		readonlyArray?: readonly (string | undefined)[];
+		readonlyTuple?: readonly ['foo' | undefined] | undefined;
+	};
+}
+
+interface FooRequired {
+	baz: string;
+	bar: {
+		function: (...args: any[]) => void;
```
Comment (sindresorhus) [inline test-d/required-deep.ts:31]:
```suggestion
		function: (...arguments: any[]) => void;
```

**Comment by sindresorhus** [comment]
CI is failing

File: source/required-deep.d.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		fontSize: number;
+		fontColor: string;
+		fontWeight: number;
+	}
+	autocomplete: boolean;
+	autosave: boolean;
+}
+```
+
+ @category Utilities
```
Comment (sindresorhus) [inline source/required-deep.d.ts:38]: https://github.com/sindresorhus/type-fest/blob/2f418dbbb6182c53cdac26dc3421f7e8806c789b/source/partial-deep.d.ts#L31-L34

**Comment by sindresorhus** [comment]
Friendly bump :)

**Comment by ilchenkoArtem** [comment]
> Friendly bump :)

soory, I am not having time now(( I am having a lot of work(

**Comment by sindresorhus** [comment]
Closing for lack of activity.

File: test-d/required-deep.ts (hunk 2 above)
Comment (6XGate) [reply test-d/required-deep.ts:31]: The term `arguments` is not allowed when TypeScript is in strict mode anymore, if it ever way.

**Pull request closed without merging**